### PR TITLE
Serialization infinity in alternative way

### PR DESCRIPTION
**Pull request by TonyXiang8787**
Fixes infinity handling in issue: https://github.com/PowerGridModel/power-grid-model/issues/320

"inf", "+inf" and "-inf" will be serialized

### CODE REVIEW

**Comment by TonyXiang8787** [comment]
@mgovers @Laurynas-Jagutis this is a suggestion on alternative way of handling infinite values in serialization. We only touch the json <-> msgpack conversion process.

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 1]
```diff
@@ -158,12 +158,39 @@ class Deserializer {
     std::vector<std::vector<ArraySpan>> msg_views_;
 
     static std::vector<char> json_to_msgpack(std::string_view json_string) {
-        nlohmann::json const json_document = nlohmann::json::parse(json_string);
+        nlohmann::json json_document = nlohmann::json::parse(json_string);
+        json_convert_inf(json_document);
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:162]: right... i **completely** forgot about the fact that msgpack is able to encode the full IEEE 754 range. apologies

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 2]
```diff
@@ -158,12 +158,39 @@ class Deserializer {
...
         std::vector<char> msgpack_data;
         nlohmann::json::to_msgpack(json_document, msgpack_data);
         return msgpack_data;
     }
 
+    static void json_convert_inf(nlohmann::json& json_document) {
+        switch (json_document.type()) {
+        case nlohmann::json::value_t::object:
+        case nlohmann::json::value_t::array:
+            for (auto& value : json_document) {
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:172]: i think you need the value iterator for object here, not the key iterator, right?

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 3]
```diff
@@ -158,12 +158,39 @@ class Deserializer {
...
+    }
+
+    static void json_string_to_inf(nlohmann::json& value) {
+        std::string const str = value.get<std::string>();
+        if (str == "inf" || str == "+inf") {
+            value = std::numeric_limits<double>::infinity();
+        }
+        if (str == "-inf") {
+            value = -std::numeric_limits<double>::infinity();
+        }
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:191]: i wonder: is nlohmann json able to handle this?

i guess if the tests pass, then yes

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/serializer.hpp [hunk 4]
```diff
@@ -210,13 +210,40 @@ class Serializer {
...
+    }
+
+    static void json_inf_to_string(nlohmann::json& value) {
+        double const v = value.get<double>();
+        if (std::isinf(v) && v > 0.0) {
+            value = "inf";
+        }
+        if (std::isinf(v) && v < 0.0) {
+            value = "-inf";
+        }
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/serializer.hpp:244]:
```suggestion
        if (std::isinf(v)) {
            value = v > 0.0 ? "inf" : "-inf;
        }
```

**Review by mgovers — COMMENTED** [review]
again, my bad for completely forgetting one of the properties of msgpack.

in addition, i did not realize that nlohmann json is able to handle infinity up til the serialization part.

i guess since i'll be OOO next week, either you or @petersalemink95 will need to communicate this to @Laurynas-Jagutis and decide on the way to go forward

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 5]
```diff
@@ -158,12 +158,39 @@ class Deserializer {
...
 
+    static void json_convert_inf(nlohmann::json& json_document) {
+        switch (json_document.type()) {
+        case nlohmann::json::value_t::object:
+        case nlohmann::json::value_t::array:
+            for (auto& value : json_document) {
+                json_convert_inf(value);
+            }
+            break;
+        case nlohmann::json::value_t::string:
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:176]: Sonar cloud

```suggestion
        using enum nlohmann::json::value_t;
        switch (json_document.type()) {
        case object:
        case array:
            for (auto& value : json_document) {
                json_convert_inf(value);
            }
            break;
        case string:
```

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/serializer.hpp [hunk 6]
```diff
@@ -210,13 +210,40 @@ class Serializer {
...
 
+    static void json_convert_inf(nlohmann::json& json_document) {
+        switch (json_document.type()) {
+        case nlohmann::json::value_t::object:
+        case nlohmann::json::value_t::array:
+            for (auto& value : json_document) {
+                json_convert_inf(value);
+            }
+            break;
+        case nlohmann::json::value_t::number_float:
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/serializer.hpp:229]: Sonar cloud

```suggestion
        using enum nlohmann::json::value_t;
        switch (json_document.type()) {
        case object:
        case array:
            for (auto& value : json_document) {
                json_convert_inf(value);
            }
            break;
        case number_float:
```

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp (hunk 5 above)
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:176]: You can't do that. `nlohmann::json::value_t` is a type alias to `nlohmann::detail::value_t`. `gcc/clang` does not allow you to `using enum` on a type alias, even it is aliasing a real enum class.

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp (hunk 3 above)
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:191]: See the links below

https://json.nlohmann.me/features/types/number_handling/

https://json.nlohmann.me/features/binary_formats/messagepack/

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp (hunk 2 above)
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:172]: By default `nlohmann::json` iterates values.

**Comment by TonyXiang8787** [comment]
@Laurynas-Jagutis this is fine. Please merge it.

**Pull request merged**